lib/src/Pos/Logic/Pure.hs

```diff
@@ -0,0 +1,314 @@
+import qualified Crypto.Hash as Crypto (hash)
+
+import           Pos.Core (StakeholderId, SoftwareVersion (..), BlockVersion (..),
+                           HeaderHash, Block, BlockHeader, GenericBlock (..),
+                           GenericBlockHeader (..), ExtraHeaderData, ExtraBodyData,
+                           BlockVersionData (..), ApplicationName, mkApplicationName,
+                           unsafeCoinPortionFromDouble, TxFeePolicy (..), SoftforkRule (..))
+import           Pos.Core.Block.Main
+import           Pos.Core.Common (ChainDifficulty (..), BlockCount (..))
+import           Pos.Core.Delegation (DlgPayload (..))
+import           Pos.Core.Ssc (SscPayload (..), SscProof (..), VssCertificatesMap (..))
+import           Pos.Core.Slotting (SlotId (..), EpochIndex (..), LocalSlotIndex (..))
+import           Pos.Core.Txp (TxPayload (..), TxProof (..))
+import           Pos.Core.Update (UpdatePayload (..), UpdateProof)
+import           Pos.Crypto.Hashing (AbstractHash (..), Hash)
+import           Pos.Crypto.Signing (PublicKey (..), Signature (..))
+import           Pos.Data.Attributes (UnparsedFields (..), Attributes (..))
+import           Pos.Merkle (MerkleTree (..), MerkleRoot (..))
+import           Pos.Util.Chrono (OldestFirst (..), NewestFirst (..))
+
+import           Pos.Logic.Types (Logic (..), KeyVal (..))
+
+-- | Serves up a single (invalid but well-formed) block and block header for
+-- any request.
+pureLogic
+    :: ( Applicative m )
+    => Logic m
+pureLogic = Logic
+    { ourStakeholderId   = stakeholderId
+    , getBlock           = \_ -> pure (Right (Just block))
+    , getBlockHeader     = \_ -> pure (Right (Just blockHeader))
+    , getBlockHeaders    = \_ _ -> pure (Right (NewestFirst (pure blockHeader)))
+    , getBlockHeaders'   = \_ _ -> pure (Right (Just (OldestFirst (pure mainBlockHeaderHash))))
+    , getTip             = pure (Right block)
+    , getTipHeader       = pure (Right blockHeader)
+    , getAdoptedBVData   = pure blockVersionData
+    , postBlockHeader    = \_ _ -> pure ()
+    , postPskHeavy       = \_ -> pure True
+    , postTx             = dummyKeyVal
+    , postUpdate         = dummyKeyVal
+    , postVote           = dummyKeyVal
+    , postSscCommitment  = dummyKeyVal
+    , postSscOpening     = dummyKeyVal
+    , postSscShares      = dummyKeyVal
+    , postSscVssCert     = dummyKeyVal
+    , recoveryInProgress = pure False
+    , securityParams     = def
+    , checkBlockHeader   = \_ -> pure True
+    , checkBlock         = \_ -> pure True
+    }
+  where
+    dummyKeyVal :: Applicative m => KeyVal key val m
+    dummyKeyVal = KeyVal
+        { toKey      = \_ -> error "dummy: can't make key"
+        , handleInv  = \_ -> pure False
+        , handleReq  = \_ -> pure Nothing
+        , handleData = \_ -> pure False
+        }
+
+stakeholderId :: StakeholderId
+stakeholderId = AbstractHash (Crypto.hash (mempty :: ByteString))
```

Comment:
I'd rather not.

1. It's called unsafe. Why is it unsafe? Should I be worried? What do I have to prove in order for its use to be safe?
2. Looking at its definition, it requires a `Bi` instance. But I don't want to use a `Bi` instance, I want to use the hash of an empty `ByteString`. Why should I go roundabout when I know what I want and can get it directly?
3. Is `Crypto.hash` unsafe?

Reply:
1. It is unsafe because you can hash a value of type `a`, but you get a `Hash` of `b`, where `b` can differ from `a`.
2. There is `hashRaw` if want to hash a `ByteString`. There is no `unsafeHashRaw`, because it wasn't needed. You can define it. But I don't see a problem in using `Bi` instance. You just need to get some constant hash.
3. Yes, it is. It doesn't have a phantom parameter at all. The safety here means that if you have a value of type `T` you get `Hash T`.

The whole point of wrappers from `Crypto.Hashing` is to avoid using functions from `cryptonite`. It makes the code less dependent on underlying crypto library. It in turn makes it clearer which parts of `cryptonite` we use and makes the code more resistant to changes in `cryptonite`.

Reply:
Probably intensive bikeshedding at this point, but `unsafeAddressHash/unsafeHashRaw` exist for a reason, I also think they should be used where possible.

Reply:
I suppose my main issue here is that I can't make an `AbstractHash` from a particular `ByteString`, I always have to go via `Bi`.

What's the point of having the second phantom type on `AbstractHash`? Does it really add any safety? I see the point of the `algo` type, but really, if I have two `Blake2b_256` hashes, why do I care *what they are the hash of*? 

In any case, wouldn't it be more correct to have

```Haskell
unsafeAbstractHash :: HashAlgorithm algo => ByteString -> AbstractHash algo anything
unsafeAbstractHash = AbstractHash . Hash.hash
```

?? As it's defined now, it's not "unsafe": the `Bi` instance is used to encode the thing so I think that means it fits your definition of "safe".

Reply:
It's called “unsafe” because it provides an `AbstractHash algo x` that wasn't necessarily created by hashing a value of type `x`. Perhaps `unsafe` isn't the best name for it.

> if I have two Blake2b_256 hashes, why do I care what type they are the hash of?

Having type-indexed hashes is good for:

* Documentation purposes (when you see a `Hash UpdateProposal` in a datatype it's immediately obvious what happens here).
* Playing type tetris (if you have a function or constructor that takes two hashes as arguments, in lots of cases those will be hashes of different types and you won't be able to mix them up).

Reply:
> Perhaps unsafe isn't the best name for it.

Almost definitely, in fact – it's not more unsafe than `decodeAbstractHash`. Perhaps the name will be changed at some point.

Reply:
At any rate, I agree with Ivan – `unsafeAbstractHash ()` or something like that is the best solution here, even if the name is unfortunate.

This said, I also agree with you regarding the fact that it's annoying that one has to jump through the hoops to simply get a hash of an empty bytestring. I don't like the whole `Raw` mess (and AFAIK it was me who introduced it, though I'm not sure). However, I would still much prefer if all calls to `Crypto.hash` were abstracted away. Generally, I think that “don't break abstractions without a good reason” is a useful heuristic, and here I don't see a reason that is good enough.